### ISSUE TITLE
Gecko failed to connect to Firefox57.0.4 when running from Bamboo agent

### ISSUE THREAD

**Issue by Eskampette** [mozilla/geckodriver#1141]
## System
* Version: geckodriver-v0.19.1-win64
* Platform: Windows Server 2012R2
* Firefox: 57.0.4 (64-bit)
* Selenium: selenium-server-standalone-3.8.1
* Language: Java8
* Integration with Bamboo

## Testcase

We are using Bamboo (Bamboo agent are running as service on our Window server box). Everything ws OK with old version of Firefox (50.0), but since we update it to the last version (57.0.4), our tests failed with the following logs (cf. [plan-27100469-PRODCHROME-59.log](https://github.com/mozilla/geckodriver/files/1653683/plan-27100469-PRODCHROME-59.log)):

> build	22-Jan-2018 16:32:04	   [testng] ###!!! [Parent][MessageChannel] Error: (msgtype=0x15008F,name=PBrowser::Msg_UpdateNativeWindowHandle) Channel error: cannot send/recv
> build	22-Jan-2018 16:32:04	   [testng] 
> build	22-Jan-2018 16:32:04	   [testng] 
> build	22-Jan-2018 16:32:04	   [testng] ###!!! [Parent][MessageChannel] Error: (msgtype=0x150083,name=PBrowser::Msg_Destroy) Channel error: cannot send/recv
> build	22-Jan-2018 16:32:04	   [testng] 
> build	22-Jan-2018 16:32:04	   [testng] 
> build	22-Jan-2018 16:32:04	   [testng] ###!!! [Parent][MessageChannel] Error: (msgtype=0x15008F,name=PBrowser::Msg_UpdateNativeWindowHandle) Channel error: cannot send/recv
> build	22-Jan-2018 16:32:04	   [testng] 
> build	22-Jan-2018 16:32:04	   [testng] 
> build	22-Jan-2018 16:32:04	   [testng] ###!!! [Parent][MessageChannel] Error: (msgtype=0x150083,name=PBrowser::Msg_Destroy) Channel error: cannot send/recv
> build	22-Jan-2018 16:32:04	   [testng] 

I also attach the Firefox profile folder: [rust_mozprofile.fmglif8m47VA.zip](https://github.com/mozilla/geckodriver/files/1653687/rust_mozprofile.fmglif8m47VA.zip)

## Stacktrace

[plan-27100469-PRODCHROME-59.log](https://github.com/mozilla/geckodriver/files/1653683/plan-27100469-PRODCHROME-59.log)

## Trace-level log

> FirefoxOptions options = new FirefoxOptions();
> options.setHeadless(false);
> options.setCapability("browser.tabs.remote.autostart", "true");
> options.setCapability("browser.tabs.remote.autostart.1", "true");
> options.setCapability("browser.tabs.remote.autostart.2", "true");
> options.setCapability("security.sandbox.content.level", "4");
> options.setLogLevel(FirefoxDriverLogLevel.TRACE);
            
> WebDriver wb = new FirefoxDriver(options);

## More information

Everything is fine when running the same code locally. The problem only happen when running our tests from Bamboo agent.

I saw similar issues but not exactly the same. I wonder if we can have more information from the log to try to identify the problem.

**Comment by whimboo** [comment]
This could be a dupe of issue #466. Can you please test with the latest Firefox 58.0 beta release which can be found here: https://www.mozilla.org/en-US/firefox/channel/desktop/#beta

Most likely this crash in Firefox should be also fixed for you.

**Comment by Eskampette** [comment]
Thanks Henrik. I am gonna try that and will let you know the result.

**Comment by Eskampette** [comment]
Indeed, seems to be the same problem than #466:
* It doesn't work with Firefox 58.0b16 or the official 58.0,
* It work with the [Firefox nightly](https://www.mozilla.org/en-US/firefox/channel/desktop/#nightly) [no need to set the "security.sandbox.content.level" capabilities on gecko driver].

Is there any solution for using the official firefox release?

**Comment by whimboo** [comment]
@Eskampette just use the same sandbox level as for Nightly (please note that it could mean some unforeseen misbehavior given that we don't use that value on purpose on releases yet), or use the Nightly version of Firefox for now. I would suggest that you do the former first.

Marking issue as dupe of #466.

**Comment by Eskampette** [comment]
@whimboo, on nightly, the "security.sandbox.content.level" is set to "5". I tried to use the same with Firefox  58.0 (or beta) but it does not work using this piece of code when using Gecko driver (Java):

> FirefoxOptions options = new FirefoxOptions();
> options.setCapability("security.sandbox.content.level", "5");
> WebDriver wb = new FirefoxDriver(options);

Is it the right thing to do?

As a side note I should say that I really don't have a clue about the meaning of the "security.sandbox.content.level" property and expected behavior with different value.

**Comment by andreastt** [comment]
`security.sandbox.content.level` is a preference, not a WebDriver capability. You will want to pass it in to the `moz:firefoxOptions` dictionary. See the README.

**Comment by Eskampette** [comment]
Thanks @andreastt. As I use Java FirefoxOptions, I think I can use the preference:

```
FirefoxOptions options = new FirefoxOptions();
options.addPreference("security.sandbox.content.level", "5");
return new FirefoxDriver(options);
```

From the log (which seems to follow the README):
>build	24-Jan-2018 13:37:15	   [testng] 2018/01/24 13:37:14.970 INFO [com.foliofn.selenium.util.webdriver.WebDriverFactory][main]: Firefox options: {acceptInsecureCerts=true, browserName=firefox, moz:firefoxOptions={args=[], prefs={security.sandbox.content.level=5}}} 

However, it still doesn't work with Firefox 58.0.

**Comment by Eskampette** [comment]
It seems that my problem was more about the way to set the "moz:firefoxOptions" preference with Java...

It work with Firefox 58.0 using the following code:

````
FirefoxOptions options = new FirefoxOptions();
options.setHeadless(true);
options.addPreference("browser.tabs.remote.autostart", false);
options.addPreference("browser.tabs.remote.autostart.2", false);
options.addPreference("security.sandbox.content.level", "5");

WebDriver wb = new FirefoxDriver(options);
````

Thanks for your support.

**Comment by whimboo** [comment]
@Eskampette out of interest why are you disabling multi-process (e10s) in Firefox? Is there anything which fails in your case?

**Comment by Eskampette** [comment]
@whimboo, we switch back to Firefox 57.0.4 (do not ask me why...) and, from our tests (*), it works only with this configuration (headless is not necessary):

````
FirefoxOptions options = new FirefoxOptions();
options.addPreference("browser.tabs.remote.autostart", false);
options.addPreference("browser.tabs.remote.autostart.2", false);
options.addPreference("security.sandbox.content.level", "5");
````

(*) Selenium tests is running on Bamboo as a window service.

**Comment by whimboo** [comment]
So what fails if you remove those two preferences? It would be great if you could check that, also because since 58 got released the 57.0.4 is no longer officially supported. A trace log of the failure would be great to have. Thanks.

**Comment by Eskampette** [comment]
@whimboo, if I remove this 2 preferences, the Gecko Driver cannot connect to Firefox: initial problem reported on this ticket.

Here are the log / profile when running my tests against Firefox 58.0 (Same selenium / gecko driver version), and with the following preferences:

```
FirefoxOptions options = new FirefoxOptions();
options.addPreference("security.sandbox.content.level", "5");
options.setLogLevel(FirefoxDriverLogLevel.TRACE);
```
[plan-27100469-PRODCHROME-116.log](https://github.com/mozilla/geckodriver/files/1667711/plan-27100469-PRODCHROME-116.log)
[rust_mozprofile.F2b5kcDaeKfe.zip](https://github.com/mozilla/geckodriver/files/1667712/rust_mozprofile.F2b5kcDaeKfe.zip)

Let me know if you need anything else.

**Comment by whimboo** [comment]
@Eskampette thanks a lot for the profile. It contained minidump files for the crashes you are experiencing when running the tests with multiple processes. Those can be found under the `minidumps` sub folder. 

I uploaded them for crash analysis and here is what you experience:
https://crash-stats.mozilla.com/report/index/b3673df2-88d7-4f13-8ff4-d0d4b0180130#tab-details

This actually guides me to https://bugzilla.mozilla.org/show_bug.cgi?id=1400637.

It might be related to issue #1134. Could you try to test with a recent Firefox 59 beta or even 60 Nightly? You can download them here: https://www.mozilla.org/en-US/firefox/channel/desktop/

Are you using any AV software which could interfere here?

**Comment by Eskampette** [comment]
**> Could you try to test with a recent Firefox 59 beta or even 60 Nightly?** 
Last time we test it works with the nightly, without the specific preferences (c.f. [this comment](https://github.com/mozilla/geckodriver/issues/1141#issuecomment-359923331)).

It doesn't work with the beta.

**> Are you using any AV software which could interfere here?**
No, I don't think so, but I will speak with our admin. team.

**Comment by whimboo** [comment]
Please note that Nightly is always special and could have a different value for `security.sandbox.content.level`. You want to check that and then apply the same level to the profile when running with a beta build.

**Comment by Eskampette** [comment]
@whimboo, do you want me to test and attach the log / profile?

**Comment by whimboo** [comment]
If you could do that, it would be great, yes.

**Comment by FranciscoSan** [comment]
Good morning, 

I have a similar problem using this system:
    Version: geckodriver-v0.19.1-win64
    Platform: Windows Server 2012R2
    Firefox: 58 (64-bit)
    Selenium: selenium-server-standalone-3.8.1
    Language: Java8
    Integration with Bamboo

It is similar to this issue but the problem is that the browser doesn´t launch in the machine. 
I run the plan in bamboo, the job starts and I see the traces:

20-Feb-2018 11:22:03	[INFO] -------------------------------------------------------
20-Feb-2018 11:22:03	[INFO]  T E S T S
20-Feb-2018 11:22:03	[INFO] -------------------------------------------------------
20-Feb-2018 11:22:04	[INFO] Running.PREnhTest
20-Feb-2018 11:22:05	DEBUG ConfigurationUtils - ConfigurationUtils.locate(): base is null, name is config.properties
20-Feb-2018 11:22:05	DEBUG ConfigurationUtils - Loading configuration from the context classpath (config.properties)
20-Feb-2018 11:22:05	DEBUG ConfigurationUtils - ConfigurationUtils.locate(): base is null, name is urls.properties
20-Feb-2018 11:22:05	DEBUG ConfigurationUtils - Loading configuration from the context classpath (urls.properties)
20-Feb-2018 11:22:06	1519122126939        geckodriver        INFO        geckodriver 0.19.1
20-Feb-2018 11:22:06	1519122126939        geckodriver        INFO        Listening on 127.0.0.1:3412
20-Feb-2018 11:22:08	1519122128001        mozrunner::runner        INFO        Running command: "C:\\Program Files\\Mozilla Firefox\\firefox.exe" "-marionette" "-profile" "C:\\Windows\\system32\\config\\systemprofile\\AppData\\Local\\Temp\\rust_mozprofile.bhUlm5SkhK2x"
20-Feb-2018 11:22:08	1519122128780        Marionette        INFO        Enabled via --marionette
20-Feb-2018 11:22:10	1519122130573        Marionette        INFO        Listening on port 61319
20-Feb-2018 11:22:11	1519122131153        Marionette        WARN        TLS certificate errors will be ignored for this session
20-Feb-2018 11:22:11	Feb 20, 2018 11:22:11 AM org.openqa.selenium.remote.ProtocolHandshake createSession
20-Feb-2018 11:22:11	INFO: Detected dialect: W3C
20-Feb-2018 11:22:11 DEBUG SeleniumTest - TestMethod: nh001Test

The testSuite start and I see the trace in the bamboo logs, Notifying me that the testcases are working, but the doubt it is that in windows Server 2012R12 the browser doesn´t launch(firefox not open). 
¿Why this happens?¿The testSuite it runnings?¿Who can I launch the browser?
The idea it is to cacth a picture when the TestCase fail.
I am using this configuration in my code:

private static FirefoxOptions firefoxOptions() {
		FirefoxOptions optionsFireFox = new FirefoxOptions();
		optionsFireFox.addPreference("browser.tabs.remote.autostart", false);
		optionsFireFox.addPreference("browser.tabs.remote.autostart.2", false);
		optionsFireFox.addPreference("security.sandbox.content.level", "5");
		return optionsFireFox;
	}

	public static WebDriver setupSelenium(String browser, String browserRC) {
	WebDriver driver=null;
        UrlProperties.leerUrlPropiedades();


        File driverFileFirefox= new File(UrlProperties.getFirefoxDriver());
        System.setProperty("webdriver.gecko.driver", driverFileFirefox.getAbsolutePath());

        String browserAux=browser;
 
		switch(browser) {
		case ConstantConfigTest.FIREFOX:
			driver=new FirefoxDriver(firefoxOptions());
		break;
		case ConstantConfigTest.CHROME:
			driver=new ChromeDriver(chromeOptions());
		break;
                  default:
                  break;
		}

Other problem it is with the chrome that give me the following error: 
Chrome not reachable 
I would like to execute the CI with bammbo remote agent and  chrome or firefox opened in windows server 2012R12 to view the testsuite executation.
Sorry to write in a closed issue.
Regard

**Comment by Eskampette** [comment]
@FranciscoSan, I am not part of the support team, but if you launch your Selenium test with Bamboo, you will not see any browser open, not matter it's Firefox / Chrome.

Indeed, Bamboo runs plan based on agents, running on your server box (i.e. windows Server 2012R12 on your case). Those agents are running as a Window service, hence they are using a special window session, with no access to local desktop. However, your browser is running internally, which allow your tests to be executed.

**From what I understood from your description, the problem is not the same than mine. So it should be threaded on on a different ticket.**

Here our piece of (Java) code to take a screenshot. You can give it a try:
````
import org.apache.commons.io.FileUtils;

private static final String SCREENSHOT_DIR = "screenshots";

public final class SeleniumUtils {

    public static final void screenshot(WebDriver wd, String destFile) {
        try {
            File srcFile = ((TakesScreenshot) wd).getScreenshotAs(OutputType.FILE);
            FileUtils.copyFile(srcFile, getFileWithTimestamp(SCREENSHOT_DIR, destFile));
        } catch (Throwable t) {
            logger.warn("Error while taking screenshot on " + destFile, t);
        }
    }
}
````

**Comment by FranciscoSan** [comment]
@Eskampette thank you so much for the answer. I did not know that functionality of  the windows service. 
I have used the code that you provide me and it  is working .
I execute the bamboo agent and browser is not launch but it take the picture of the error detected. 
So with the explanation of the Window service functionality that you give me, the function is considered correct with the system that I am using.
Thank so much.
Regard
